docs/rfcs/1005-realm-key-rotation.md

```diff
@@ -0,0 +1,821 @@
+
+> **Note:**
+>
+> - We cannot simply get the keys-bundle that originally created a given key given
+>   the realm might have been shared with the user later on.
+> - A more powerful algorithm would be to combine the valid keys in each keys-bundles.
+>   (e.g. we can self-heal if bundle A contains key X corrupted and bundle B contains
+>   key Y corrupted). In practice however it is likely a bundle is either totally
+>   valid or totally broken.
+
+### 6.4 Automatic key rotation
+
+For automatic key rotation, a new "key rotation manager" is to be introduced. It role
+is very simple:
+
+1. Wait for new `UserRevokedCertificate`, `RealmRoleCertificate` and
+   `RealmKeyRotationCertificate` certificates
+2. For `UserRevokedCertificate`, call `realm_unshare` for each realm we are owner (and
+   the revoked user is part of)
+3. For `RealmRoleCertificate`, if the role is None and ware are owner of the realm:
+  3.1. Wait a random time between 30 and 60 seconds (and reset the wait if we were
+       already waiting for this realm)
+  3.2. Call `realm_rotate_key`
+4. For `RealmKeyRotationCertificate` cancel the wait from step 3.1 if any
+
+The wait at step 3.1 is here for two reasons:
+
+- Give priority to the author of the certificates to do the key rotation. This is
+  typically needed if he is doing un-sharing batch (and hence will do the key rotation
+  only once all un-sharing have been done)
+- Avoid all owners to send key rotation command at the same time to the server.
+
+### 6.3 Existing local data: `LocalUserManifest`
+
+`LocalUserManifest` has to be modified to reflect changes in `UserManifest`:
+
+- remove `last_processed_message` field
+- change `workspace` field to use the updated `WorkspaceEntry` type
+
+## 7 Backward compatibility & migration
+
+As stated, we don't support APIv3 and APIv4 in the same Parsec server.
+
+An APIv3 Parsec server can be migrate to APIv4.
+
+After that it will deny any APIv3 request, hence forcing all it users to upgrade
+they Parsec client at once.
+
+### 7.1 Server-side organization migration
+
+This is typically for the SAAS version Parsec: some clients may want to be able
+to control when the switch from v2 to v3 should come for they organization.
+
+The solution for that is to run create a new separate Parsec instance running on v3:
+
+- All existing organizations stay on the legacy Parsec v2 instance
+- All new organizations are created on the new Parsec v3 instance
+- When organization owner requests it, his organization can be migrated from
+  legacy to the new Parsec v3 instance.
+
+The migration goes as follow:
+
+1) A partial database export is done from the Parsec v2 instance containing only the
+  organization(s) to migrate
+2) The partial export is a valid Parsec v2 database on it own. Hence the Parsec v2 -> v3
+   database migration script can be run on it.
+3) Each organization of the migrated database is inserted in the Parsec v3 instance
```

Comment:
the organization must be disabled in the Parsec v2 server once migration is done, this is to avoid Parsec v2 client to access the old data (instead of parsec v3 server)